curvlinops/kfac.py

```diff
@@ -114,6 +115,14 @@ def __init__(
                 from the parameters. Defaults to ``None``.
             seed: The seed for the random number generator used to draw labels
                 from the model's predictive distribution. Defaults to ``2147483647``.
+            fisher_type: The type of Fisher/GGN to approximate. If ``'type-2'``, the
+                expectation over the model outputs is computed exactly by
+                computing the backwward pass for each output dimension. This is
+                sometimes also called type-2 Fisher. If ``'mc'``, the expectation
```

Comment:
This description is correct for `MSELoss`, but not for `CrossEntropyLoss`.
- For `MSELoss`, the expected Hessian of the loss w.r.t. the model's output under the model's distribution is `\nabla_f^2 \ell \propto I`. Then, we take each column `c_i \propto e_i`  of `I`, which is a one-hot vector, and compute  `\nabla (f^\top c_i)` for all values of `i`. Hence, this is like calling `.backward` on `f[0], f[1], ...`.
- For `CrossEntropyLoss`, the expected Hessian is `\nabla_f^2 \ell = diag(p) - p p^T` where `p = softmax(f)`. In this case, taking a column does not yield a one-hot vector. Hence, this is not calling `.backward` on `f[0], f[1], ...`. It is calling `backward` on `(\nabla_f^2 \ell)[:, i]` for `i=0, 1, ...`.

Reply:
I see now that the wording is ambiguous. I meant to simply describe the overhead by stating the number of backward passes that is required for the selected `fisher_type`. In both cases, for the `MSELoss` and `CrossEntropyLoss`, this is equal to the output dimension/the number of classes for `'type-2'`. I did not want to add any details on the maths or the specific implementation.

Is this clearer?

> If ``'type-2'``, the exact Hessian of the loss w.r.t. the model outputs is used. This requires as many backward passes as the output dimension, i.e. the number of classes for classification.

Reply:
Yes, that's great!